### PR TITLE
All channels

### PR DESCRIPTION
**Pull request by valeriadamante**
This PR is intended to include all the other channels

### CODE REVIEW

File: config/Run2_2016/triggers.yaml [hunk 1]
```diff
@@ -1,10 +1,20 @@
...
+# 0 => LooseIso, 1 => Medium(Comb)Iso, 2 => VLooseIso, 3 => None, 4 => L2p5 pixel iso, 5 => OverlapFilter IsoMu, 6 => OverlapFilter IsoEle, 7 => L1-HLT matched, 8 => Dz for Tau;
+
 ditau:
   channels:
     - tauTau
   path_data:
-    - HLT_DoubleMediumCombinedIsoPFTau35_Trk1_eta2p1_Reg && run >= 278820
-    - HLT_DoubleMediumIsoPFTau35_Trk1_eta2p1_Reg && run < 278820
+    - HLT_DoubleMediumCombinedIsoPFTau35_Trk1_eta2p1_Reg && isEraH # && run >= 278820
```
Comment (kandrosov) [inline config/Run2_2016/triggers.yaml:16]: What is the issue with the run number?

File: config/Run2_2018/errorsQCD.txt
Comment (kandrosov) [inline config/Run2_2018/errorsQCD.txt:1]: remove from PR

File: include/Utilities.h [hunk 2]
```diff
@@ -22,6 +22,7 @@ namespace detail {
       return shifted - central;
     }
     static T FromDelta(const T& delta, const T& central){
+      // std::cout << "delta = " << delta << " central = " << central << std::endl;
```
Comment (kandrosov) [inline include/Utilities.h:25]:
```suggestion
```

File: include/Utilities.h [hunk 3]
```diff
@@ -40,7 +41,9 @@ namespace detail {
     static ROOT::VecOps::RVec<T> FromDelta(const ROOT::VecOps::RVec<T>& delta, const ROOT::VecOps::RVec<T>& central){
       ROOT::VecOps::RVec<T> fromDeltaVec = delta;
       size_t n_max = std::min(delta.size(), central.size());
+      // std::cout<< "delta vec size = "<< fromDeltaVec.size() <<  " central size = "<<central.size()<<std::endl;
```
Comment (kandrosov) [inline include/Utilities.h:44]:
```suggestion
```

File: include/Utilities.h [hunk 4]
```diff
@@ -40,7 +41,9 @@ namespace detail {
...
       for (size_t n =0 ; n < n_max; ++n){
+        // std::cout<<"deltaVec["<<n<<"] = "<<fromDeltaVec[n]<< " central["<<n<<"] = "<< central[n]<<std::endl;
```
Comment (kandrosov) [inline include/Utilities.h:46]:
```suggestion
```

File: include/pnetSF.h
Comment (kandrosov) [inline include/pnetSF.h:1]: move this file to corrections

File: MergeCommand.sh
Comment (kandrosov) [inline MergeCommand.sh:1]: remove from PR

File: Common/Setup.py [hunk 5]
```diff
@@ -74,9 +74,11 @@ def apply_customisations(config_dict, customisations):
...
         cfg_entry = config_dict
         for key in key_entries[:-1]:
-            cfg_entry = cfg_entry[key]
-        entry_type = type(cfg_entry[key_entries[-1]])
-        cfg_entry[key_entries[-1]] = entry_type(value)
+            if key in config_dict.keys():
+                cfg_entry = cfg_entry[key]
+        if key_entries[-1] in config_dict.keys():
+            entry_type = type(cfg_entry[key_entries[-1]])
+            cfg_entry[key_entries[-1]] = entry_type(value)
```
Comment (kandrosov) [inline Common/Setup.py:81]: Undo modifications: if customization is not applied, the code shouldn't silently ignore it.

File: Analysis/tasks.py [hunk 6]
```diff
@@ -123,58 +134,81 @@ def create_branch_map(self):
...
         sample_name, prod_br, var, need_cache = self.branch_data
+        isTTCR,isDYCR,is2p5,isSC=getCustomisations(self.customisations,self.version)
         outFileName = os.path.basename(self.input()[0].path)
-        output_path = os.path.join(self.version, self.period, 'prod', var, f'{sample_name}_{outFileName}')
+        prod_dir = 'prod_SR'
+        if isDYCR:
+            prod_dir = 'prod_DYCR'
+        if isTTCR:
+            prod_dir = 'prod_TTCR'
+        output_path = os.path.join(self.version, self.period, prod_dir, var, f'{sample_name}_{outFileName}')
```
Comment (kandrosov) [inline Analysis/tasks.py:144]: why do you need this mechanism? can you simply run it with different version?

File: Analysis/tasks.py [hunk 7]
```diff
@@ -123,58 +134,81 @@ def create_branch_map(self):
...
                     ps_call(HistProducerFile_cmd, verbose=1)
             else:
                 ps_call(HistProducerFile_cmd, verbose=1)
 
+
+
 class HistProducerSampleTask(Task, HTCondorWorkflow, law.LocalWorkflow):
     max_runtime = copy_param(HTCondorWorkflow.max_runtime, 2.0)
     n_cpus = copy_param(HTCondorWorkflow.n_cpus, 1)
+    use_FileTask = luigi.BoolParameter(default=True)
```
Comment (kandrosov) [inline Analysis/tasks.py:193]: What is use-cases when fileTask is not needed? I propose to undo this modification

File: AnaProd/tasks.py [hunk 8]
```diff
@@ -167,11 +167,14 @@ def run(self):
                 anatuple_cmd = [ 'python3', producer_anatuples, '--period', self.period,
                                  '--inFile', local_input.path, '--outDir', outdir_anatuples, '--sample', sample_name,
                                  '--anaTupleDef', anaTupleDef, '--anaCache', anaCache_input.path ]
+                isCC = 'CC' in self.version.split('_')
```
Comment (kandrosov) [inline AnaProd/tasks.py:170]: undo this logic

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Analysis/tasks.py (hunk 7 above)
Comment (valeriadamante) [reply Analysis/tasks.py:193]: If you remove the histograms from histProducerFileTask to save space!

File: Common/Utilities.py [hunk 9]
```diff
@@ -131,9 +186,14 @@ def CreateColumnTypes(self):
...
         colNames[lumiIdx], colNames[3] = colNames[3], colNames[lumiIdx]
         self.colNames = colNames
+        self.colTypes = [str(self.df.GetColumnType(c)) for c in self.colNames]
+        # for n,c in zip(self.colNames,self.colTypes):
+        #      print(n,c)
+        #      self.df.Filter("event==16677").Display({n}).Print()
+        # print(self.colNames)
+        # print(self.colTypes)
         #if "kinFit_result" in self.colNames:
         #    self.colNames.remove("kinFit_result")
```
Comment (kandrosov) [inline Common/Utilities.py:196]:
```suggestion
```

File: Common/Utilities.py [hunk 10]
```diff
@@ -149,13 +209,19 @@ def CreateFromDelta(self,central_columns,central_col_types):
                 continue
             var_name_forDelta = var_name.removesuffix("Diff")
             central_col_idx = central_columns.index(var_name_forDelta)
+            #print(var_name_forDelta, var_name, central_col_idx, central_columns[central_col_idx])
```
Comment (kandrosov) [inline Common/Utilities.py:212]:
```suggestion
```

File: Common/Utilities.py [hunk 11]
```diff
@@ -149,13 +209,19 @@ def CreateFromDelta(self,central_columns,central_col_types):
...
             if central_columns[central_col_idx]!=var_name_forDelta:
                 raise RuntimeError(f"CreateFromDelta: {central_columns[central_col_idx]} != {var_name_forDelta}")
             self.df = self.df.Define(f"{var_name_forDelta}", f"""analysis::FromDelta({var_name},
                                      analysis::GetEntriesMap()[std::make_tuple(entryIndex, run, event, luminosityBlock)]->GetValue<{self.colTypes[var_idx]}>({central_col_idx}) )""")
+            #self.df.Filter("event==16677").Display({var_name_forDelta}).Print()
```
Comment (kandrosov) [inline Common/Utilities.py:217]:
```suggestion
```

File: Common/Utilities.py [hunk 12]
```diff
@@ -149,13 +209,19 @@ def CreateFromDelta(self,central_columns,central_col_types):
...
             var_list.append(f"{var_name_forDelta}")
+
+    def AddMissingColumns(self,central_columns,central_col_types):
         for central_col_idx,central_col in enumerate(central_columns):
-            if central_col in var_list or central_col in self.colNames: continue
+            #print(central_col)
+            #print(central_col in self.colNames)
```
Comment (kandrosov) [inline Common/Utilities.py:223]:
```suggestion
```

File: AnaProd/tasks.py [hunk 13]
```diff
@@ -155,6 +178,10 @@ def run(self):
         thread = threading.Thread(target=update_kinit_thread)
         thread.start()
         anaCache_remote = self.input()[0]
+        customisation_dict = getCustomisationSplit(self.customisations)
+        channels = customisation_dict['channels'] if 'channels' in customisation_dict.keys() else self.global_params['channelSelection']['signal']
```
Comment (kandrosov) [inline AnaProd/tasks.py:182]: Don't split `channelSelection` into `signal` and `control`.

**Review by kandrosov — CHANGES_REQUESTED** [review]

File: Common/Utilities.py [hunk 14]
```diff
@@ -92,6 +137,16 @@ def generate_enum_class(cls):
...
+def checkLists(list1, list2):
+    if len(list1) != len(list2):
+        print(f"lists have different length: {list1} and {list2}")
+        return False
+    for item in list1:
+        if item not in list2:
+            print(f"{item} in {list1} but not in {list2}")
+            return False
+    return True
+
```
Comment (kandrosov) [inline Common/Utilities.py:149]:
```suggestion
```

File: Common/Utilities.py [hunk 15]
```diff
@@ -47,6 +47,51 @@ class WorkingPointsbTag(Enum):
     Medium = 2
     Tight = 3
 
+WorkingPointsParticleNet = {
```
Comment (kandrosov) [inline Common/Utilities.py:50]: move to corrections

File: run_tools/law_customizations.py [hunk 16]
```diff
@@ -18,18 +18,26 @@ def copy_param(ref_param, new_default):
...
     """
     Base task that we use to force a version parameter on all inheriting tasks, and that provides
     some convenience methods to create local file and directory targets at the default data path.
     """
-
     version = luigi.Parameter()
+    prefer_params_cli = [ 'version' ]
     period = luigi.Parameter()
     customisations =luigi.Parameter(default="")
+    consider_former_tasks = luigi.BoolParameter(default=True)
```
Comment (kandrosov) [inline run_tools/law_customizations.py:40]:
```suggestion
```

File: AnaProd/tasks.py [hunk 17]
```diff
@@ -107,8 +122,14 @@ class AnaTupleTask(Task, HTCondorWorkflow, law.LocalWorkflow):
     max_runtime = copy_param(HTCondorWorkflow.max_runtime, 40.0)
     n_cpus = copy_param(HTCondorWorkflow.n_cpus, 4)
 
+    def GetVersions(self):
+        AnaCacheTask_version = get_param_value(AnaCacheTask, 'version') if get_param_value(AnaCacheTask, 'version') else self.version
+        InputFileTask_version = get_param_value(InputFileTask, 'version') if get_param_value(InputFileTask, 'version') else self.version
+        return AnaCacheTask_version,InputFileTask_version
```
Comment (kandrosov) [inline AnaProd/tasks.py:128]:
```suggestion
```

File: AnaProd/tasks.py [hunk 18]
```diff
@@ -210,20 +239,29 @@ def run(self):
 class DataMergeTask(Task, HTCondorWorkflow, law.LocalWorkflow):
     max_runtime = copy_param(HTCondorWorkflow.max_runtime, 5.0)
 
+    def GetVersions(self):
+        AnaTupleTask_version = get_param_value(AnaTupleTask, 'version') if get_param_value(AnaTupleTask, 'version') else self.version
+        AnaCacheTask_version = get_param_value(AnaCacheTask, 'version') if get_param_value(AnaCacheTask, 'version') else self.version
+        InputFileTask_version = get_param_value(InputFileTask, 'version') if get_param_value(InputFileTask, 'version') else self.version
+        return AnaTupleTask_version,AnaCacheTask_version,InputFileTask_version
```
Comment (kandrosov) [inline AnaProd/tasks.py:246]:
```suggestion
```

**Review by kandrosov — CHANGES_REQUESTED** [review]

**Review by kandrosov — APPROVED** [review]

**Pull request merged**